### PR TITLE
Create `isSubypeQualifiers()` and `isSubtypeShallow()` methods

### PR DESCRIPTION
**Pull request by mernst**
Merge together:
 * https://github.com/typetools/checker-framework/pull/5862
 * https://github.com/plume-lib/multi-version-control/pull/224
 * https://github.com/plume-lib/plume-util/pull/304
 * https://github.com/codespecs/daikon/pull/478

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:2392]: It's not clear how this is different from `isRelevant`.

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:2366]: It just seems like this (and the other calls to `canBeAnnotated`)  is a workaround for the defaulting rules in the Signedness Checker that aren't working correctly.

**Review by smillst — CHANGES_REQUESTED** [review]

**Comment by mernst** [comment]
@smillst CI has passed on all jobs, on one JDK.  Thanks for the fix.

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:2392]: It isn't any more.  I actually have a branch that removes it; I'm waiting for CI to run on that branch.

File: checker/src/main/java/org/checkerframework/checker/signedness/SignednessChecker.java
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/signedness/SignednessChecker.java:15]: So this makes it so there will be (almost) no warnings about `char`s.  I don't understand why this is safe.

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:2972]: Here you check `isCompoundType`, but in other places you don't.

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:2366]: How did you decide which checks to skip if a type isn't relevant? Would is be better to change isSubtype instead?

Comment (mernst) [reply framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Most calls to `isSubtype()` are to the version in `QualifierHierarchy`, which takes `AnnotationMirror`s as arguments.  That `isSubtype()` cannot behave differently based on the Java basetype.  So, the relevance test must be done before `QualifierHierarchy.isSubtype()` is called.  I put tests in places that seemed right, based on examining all occurrences of `qualifierHierarchy.isSubtype()`.

To improve uniformity and reduce the chance of errors, we could do one of these:
 * Pass the `AnnotatedTypeMirror`s into `QualifierHierarchy.isSubtype()` as an extra argument.
 * Discourage/prevent clients from calling `QualifierHierarchy.isSubtype()`, as in `myTypeFactory.getQualifierHierarchy().isSubtype()`.  Instead, have a method `AnnotatedTypeFactory.isSubtype()` that takes `AnnotatedTypeMirror`s, checks relevance, and then delegates to the `QualifierHierarchy`.
 * Keep it as is, and prevent/fix problems through code review and testing.
 * Other ideas?

All that said, I did change the `isSubtype()` that takes annotated types as well.

File: checker/src/main/java/org/checkerframework/checker/signedness/SignednessChecker.java
Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/SignednessChecker.java:15]: Since the programmer cannot annotate `char`s, there isn't much that can go wrong.
An assignment like
```java
  int i = myChar;
```
is fine regardless of the signedness of `i`.
A narrowing like
```java
  byte b = (byte) myChar;
```
is also fine regardless of the signedness of `b`.

There used to be tests that used signed characters, but that is not sensible and I removed those tests.  The checker does issue an error if the programmer writes `@Signed char` as a type.

**Comment by mernst** [comment]
> A few of the tests are crashing.

Right, I [know](https://github.com/typetools/checker-framework/pull/5862#issue-1708627530).  I don't know why, though, and could use your help in diagnosing if possible.

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:2377]: This is now the only use of `alwaysRelevantTM`.  I think we should just delete it.

File: framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java:288]: This doesn't use the passed type mirrors.  Does the following make sense?
```suggestion
    boolean tm1IsRelevant = atypeFactory.isRelevant(tm1);
    boolean tm2IsRelevant = atypeFactory.isRelevant(tm2);
    if (tm1IsRelevant == tm2IsRelevant) {
      return leastUpperBoundQualifiers(qualifier1, qualifier2);
    } else if (tm1IsRelevant) {
      return qualifier1;
    } else { // if(tm2IsRelevant) {
      return qualifier2;
    }
```

File: framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java:168]: This is only used in CFAbstractValue.  The other places that I thought it would be useful had different semantics.  So, it might make sense to move this method there.

File: framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java:345]: I'm wondering if the `isSubtypeShallowEffective` methods would make more sense in the `TypeHierarchy` class?  Also, maybe there's a better name?

**Review by smillst — COMMENTED** [review]
@mernst  This is ready for review.  I've left two comments that might make more sense to discuss at our meeting tomorrow.

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java:344]: I'm unclear why to call `lowestQualifier` if we already know that `a` is a (shallow) subtype of `b`.  That is, is there something more specific we can do when we have that knowledge?

I see that `lowestQualifier` breaks nondeterminism, but does that nondeterminism (no unique lowest type) indicate a problem in the Checker Framework?  When does it occur?  Documentation here and/or in `lowestQualifier` would help me to understand.

**Review by mernst — COMMENTED** [review]
One quick comment, while I'm working on other parts of the code review.

File: framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java:166]: This comment says "effective", but the code uses "getPrimaryAnnotations".

File: framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java:186]: What is the "S" in "subSType" (etc.) a mnemonic for?  Oh, "shallow".  I would write it out, or use the name `subShallow` (etc.) instead.

File: framework/src/main/java/org/checkerframework/framework/type/TypeHierarchy.java
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/TypeHierarchy.java:72]: What are "other annotations" in this context?
Should this sentence be deleted?

**Comment by mernst** [comment]
@smillst Looks good, thanks.  I'm assigning this back to you because the pull request requires your approval in order to merge.

**Review by smillst — DISMISSED** [review]

**Review by smillst — APPROVED** [review]

**Pull request merged**
